### PR TITLE
Sz/days every week

### PR DESCRIPTION
**Pull request by shachar058**
Here we add a feature which allows an additional kind of recurrence sequence when creating encounter events. At event creation the user can now choose specific days of the week on which he would like to have encounters.  An encounter will then be created every week on each of the specified days until the end date specified in the encounter creation.

The created encounters can be seen in the calendar, flow board, patient file, and appointments report. 

If the user chooses a 'start date' which does not fall out on one of the specified days, the start date will automatically  change to the date on which the first encounter will take place, which will be the closest next date [to the start date] on which one of the chosen days will fall. Meaning - an encounter will NOT be created on the chosen 'start date' if it does not fall out on one of the specified days of the week.

We have finished developing this feature, and are waiting anxiously for your review!
Thank you

### CODE REVIEW

**Comment by bradymiller** [comment]
Just to clarify. Do you mean appointments rather than encounters?

**Comment by sharonco-matrix** [comment]
Yes recurrent appointments :)

**Comment by bradymiller** [comment]
Will be good to get input here from @epsdky

File: interface/main/calendar/add_edit_event.php [hunk 1]
```diff
@@ -161,16 +161,34 @@ function DOBandEncounter()
...
+ occurrence of one of these set days. */
+ function setEventDate($start_date, $recurrence){
+     $timestamp = strtotime($start_date);
+     $day = date('w', $timestamp);
+     //If the 'start date' is one of the set days
+     if(in_array(($day+1), explode(',',$recurrence))){
+         return $start_date;
+     }
+     //else: (we need to change start date to first occurrence of one of the set days)
+     $new_date = getNextEncounter($start_date, $recurrence);
```
Comment (bradymiller) [inline interface/main/calendar/add_edit_event.php]: The use of "encounter" is throwing me off. Are you just getting the next day for an appointment. If that is case, would change this or will confuse developers down the road.

File: interface/main/calendar/add_edit_event.php [hunk 2]
```diff
@@ -1392,7 +1482,34 @@ function find_available(extra) {
...
+      function isRegularRepeat($repeat){
+          if($repeat == 1 || $repeat == 2){
+              return true;
+          }
+          else{
+              return false;
+          }
+      }
+      ?>
+      <!-----------
```
Comment (bradymiller) [inline interface/main/calendar/add_edit_event.php]: recommend placing this comment within php tags instead

File: interface/main/calendar/add_edit_event.php [hunk 3]
```diff
@@ -1428,6 +1545,33 @@ function find_available(extra) {
...
+
+<tr id="days_every_week_row">
+    <td></td>
+    <td></td>
+    <td><input  type='checkbox' name='days_every_week' onclick='set_days_every_week()' <?php if (isDaysEveryWeek($repeats)) echo " checked" ?>/></td>
+    <td id="days_label"><?php echo xlt('Days Of Week:'); ?></td>
+    <td id="days">
+        <?php
+        foreach (array(1 => xl('Su') , 2 => xl('M'), 3 => xl('Tu'), 4 => xl('W'),
+                     5 => xl('Th'), 6 => xl('F'), 7 => xl('Sa')) as $key => $value)
```
Comment (bradymiller) [inline interface/main/calendar/add_edit_event.php]: would change constants to:
``````
xl('Su{{Sunday}}')
etc...

Comment (bradymiller) [inline interface/main/calendar/add_edit_event.php]: Note these are basically a comment to the translator and these stuff within the brackets are not shown when displayed.

File: interface/main/calendar/modules/PostCalendar/pnuserapi.php [hunk 4]
```diff
@@ -1487,6 +1487,53 @@ function calculateEvents($days,$events,$viewtype) {
...
+						//echo "begin printing blocks for $eventD<br />";
+						//print_r($days[$occurance]['blocks']);
+						//echo "end printing blocks<br />";
+					}
+				}
+
+				$occurance =& __increment($nd,$nm,$ny,$rfreq,$rtype);
+				list($ny,$nm,$nd) = explode('-',$occurance);
+			}
+			break;
```
Comment (bradymiller) [inline interface/main/calendar/modules/PostCalendar/pnuserapi.php]: when working in a script, maintain the formatting of that script. And generally best not to use tabs and instead use spaces.

File: library/appointments.inc.php [hunk 5]
```diff
@@ -259,6 +259,53 @@ function fetchEvents( $from_date, $to_date, $where_param = null, $orderby_param
...
+					  }
+					  //////
+				  }
+
+				  $occurance =& __increment($nd,$nm,$ny,$rfreq,$rtype);
+				  list($ny,$nm,$nd) = explode('-',$occurance);
+
+			  }
+
+			  break;
```
Comment (bradymiller) [inline library/appointments.inc.php]: when working in a script, maintain the formatting of that script. And generally best not to use tabs and instead use spaces.

**Comment by bradymiller** [comment]
did a quick review. very involved code which appears solid. will await expertise of epsdky :)

File: interface/main/calendar/add_edit_event.php (hunk 1 above)
Comment (shachar058) [inline interface/main/calendar/add_edit_event.php]: Done! Though I should mention that there are many references to "appointments" in this file which are written in the code as "encounters". That's actually where I got the idea of calling it 'encounters'...

File: interface/main/calendar/add_edit_event.php (hunk 2 above)
Comment (shachar058) [inline interface/main/calendar/add_edit_event.php]: Done

File: interface/main/calendar/add_edit_event.php (hunk 3 above)
Comment (shachar058) [inline interface/main/calendar/add_edit_event.php]: Done

File: interface/main/calendar/modules/PostCalendar/pnuserapi.php (hunk 4 above)
Comment (shachar058) [inline interface/main/calendar/modules/PostCalendar/pnuserapi.php]: Made it like the others, hope I understood what you meant. If it's not fixed yet, correct me.

File: library/appointments.inc.php (hunk 5 above)
Comment (shachar058) [inline library/appointments.inc.php]: Made it like the others, hope I understood what you meant. If it's not fixed yet, correct me.
 Add a line note

**Comment by shachar058** [comment]
Ok ready for epsdky's review! Thank you.

**Comment by sharonco-matrix** [comment]
Any news on this ?

File: interface/main/calendar/add_edit_event.php [hunk 6]
```diff
@@ -1428,6 +1549,33 @@ function find_available(extra) {
...
+        #days_every_week_row input[type="checkbox"]{float:right;}
+        #days_every_week_row div{display: inline-block; text-align: center; width: 12%;}
+        #days_every_week_row div input{width: 100%;}
+    </style>
+
+<tr id="days_every_week_row">
+    <td></td>
+    <td></td>
+    <td><input  type='checkbox' name='days_every_week' onclick='set_days_every_week()' <?php if (isDaysEveryWeek($repeats)) echo " checked" ?>/></td>
+    <td id="days_label"><?php echo xlt('Days Of Week:'); ?></td>
```
Comment (bradymiller) [inline interface/main/calendar/add_edit_event.php]: remove colon and whitespace from the constant:
``````
<td id="days_label"><?php echo xlt('Days Of Week:'); ?></td>

File: interface/main/calendar/add_edit_event.php [hunk 7]
```diff
@@ -1582,14 +1731,32 @@ function find_available(extra) {
...
-    if (f.form_repeat.checked &&
+    if ((f.form_repeat.checked || f.days_every_week.checked) &&
         (! f.form_enddate.value || f.form_enddate.value < f.form_date.value)) {
         alert('<?php echo addslashes(xl("An end date later than the start date is required for repeated events!")); ?>');
         return false;
     }
+
+    //Make sure if days_every_week is checked that at least one weekday is checked.
+    if(f.days_every_week.checked && !are_days_checked()){
+        alert('<?php echo addslashes(xl("Must choose at least one day!")); ?>');
```
Comment (bradymiller) [inline interface/main/calendar/add_edit_event.php]: here's a clearer short-cut function for this (xls combines xl and addslashes):
``````
alert('<?php echo xls("Must choose at least one day!"); ?>');

**Comment by bradymiller** [comment]
Although the code makes my head spin, don't see any big red flags. Probably the best thing I can do while await the recurrent appt expert is to test it, which will do now.

Also, since the Recurrent Appointment widget is now in the main codebase, recommend adding code to support this.

**Comment by bradymiller** [comment]
Is testing well. Will need to make the patient tracker popup larger and/or add scrollbars=1:
https://github.com/openemr/openemr/blob/master/interface/patient_tracker/patient_tracker.php#L90

File: interface/main/calendar/add_edit_event.php (hunk 6 above)
Comment (shachar058) [inline interface/main/calendar/add_edit_event.php]: done

File: interface/main/calendar/add_edit_event.php (hunk 7 above)
Comment (shachar058) [inline interface/main/calendar/add_edit_event.php]: done

**Comment by shachar058** [comment]
Finished fixing up what you pointed out. 
Waiting for epsdky's reveiw.
Thank you!

**Comment by epsdky** [comment]
Problem with week days code and  Fees > Payment...

Set an rt3 appointment prior to the current day which includes a date that corresponds to the current day date and for Payment select Cash (Payment Method), Self (Patient Coverage), 10 (Paying) press "Generate Invoice" and OK after popup ("Are you sure to post for today?") the status for the rt3 is changed to '@' so all corresponding appointment dates have this status. What should have happened is a new rt0 appointment should have been created with status '@' and original rt3 exdate updated with current day date.

(Still looking...)

**Comment by shachar058** [comment]
Hi epsdky,
I set an rt3 appointment (I take it you mean the new recurrence type with weekdays) starting from a few days ago, and the appointments are on every day of the week. I then went to payments and submitted a payment with Self + Cash  of 10 for today. It asked me "Are you sure etc." and clicked yes.
It then successfully updated the @ status only for today. I checked in the database and it created a new rt0 and added to exdate of recurrence. I couldn't recreate the problem.
Did I understand you correctly?
Waiting to hear,
shahar

**Comment by epsdky** [comment]
Hello Shahar,

Your code is not at fault. I still get the problem I described and I also get your result for the exact days you selected. If an rt3 appointment (your new recurrtype) is/was set on a date prior to the current day and the days selected led to the situation where the value for pc_eventDate is the same as the current day date, since the day that the appointment was set on was not checked and the earliest date for the checked days happened to be the same as the current day date then effectively the appointment is the same as if it had been set on the current day. In Fees > Payment there is a bug for the process of paying where the current day appointment (with no encounter) is a recurring appointment that has been set on the current date that causes the status of the recurring appointment to be changed to '@' instead of a non recurring appointment being created with '@' and original recurring appt exdate being updated.

Regards,

epsdky.

**Comment by shachar058** [comment]
@bradymiller 
Hey Brady,
As epsdky mentioned above, there is a general bug in the payment process when it comes to recurrent appointments. 
Because it is a bug that isn't connected to the new feature, is it possible to add the new feature to the code base and concurrently open up a bug in the system? It would be very good also because we can then update the new recurrence widget to support this new type of recurrence.

**Comment by bradymiller** [comment]
Sounds like a solid plan.

Still need to make the patient tracker popup larger and/or add scrollbars=1:
https://github.com/openemr/openemr/blob/master/interface/patient_tracker/patient_tracker.php#L90

@epsdky, Any more to add on this review?

**Comment by bradymiller** [comment]
btw, will place this bug on the Issues.

**Comment by shachar058** [comment]
Hey brady, 
I made the tracker popup larger.
Waiting to see in the code base!
Thanks

**Comment by epsdky** [comment]
Hello Brady,

I have looked at the new code in relation to the (active) locations in the codebase where recurring appointments are calculated and the additions to encounter_events.php that are at the center of the new recurrtype and can see no problems.

**Comment by epsdky** [comment]
Hello Shahar,

Is it necessary to duplicate the rt1 code in fetchEvents and  calculateEvents to become the rt3 code instead of adding...

~~~~~~
case '3' : 
~~~~~~

or

~~~~~~
case REPEAT_DAYS  :
~~~~~~

immediately prior to the rt1 case statements?

**Comment by epsdky** [comment]
Minor problem the following is missing from between line 402 and line 403 in add_edit_event.php...

~~~~~~
$args['days_every_week'] = "0";
~~~~~~

line 402-404 should look like...

~~~~~~
$args['form_repeat'] = "0";
$args['days_every_week'] = "0";
$args['recurrspec'] = $noRecurrspec;
~~~~~~

**Comment by shachar058** [comment]
Ok done with the above 2 fixes.
Waiting to hear.
Thanks,
Shahar

**Pull request merged**